### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by Aditya-debug15**

### CODE REVIEW

File: experiment/pretest.json [hunk 1]
```diff
@@ -2,38 +2,90 @@
...
-        "b": "Explanation 2",
-        "c": "Explanation 2",
-        "d": "Explanation 2"
+        "d": "The best case time complexity of Bellman-Ford is O(E) and the worst case time complexity is O(n^3). In the best case we would run the algorithm once for all the edges thus it will be O(E). In the worst case we will run the algorithm n times for all the edges and the maximum number of edges that can be present in a graph is n(n-1)/2. Thus the worst case time complexity is O(n^3)"
+      },
+      "correctAnswer": "d",
+      "difficulty": "intermediate"
+    },
+    {
+      "question": "Q3. Bellman-Ford is type of which algorithm?",
```
Comment (raj-vlabs) [inline experiment/pretest.json]: Please consider rephrasing this. It is not capturing the intent of the question.

File: experiment/pretest.json [hunk 2]
```diff
@@ -2,38 +2,90 @@
...
+      "answers": {
+        "a": "Greedy algorithm",
+        "b": "Dynamic programming algorithm",
+        "c": "Backtracking algorithm",
+        "d": "Divide and conquer algorithm"
+      },
+      "explanations": {
+        "b": "Bellman-Ford is a dynamic programming algorithm, because we are using the previous results to calculate the current result"
```
Comment (raj-vlabs) [inline experiment/pretest.json]: Dynamic programming is characterised by Overlapping Subproblems and Optimal Substructure. Try to bring in those terms in the explanation. Previous results are used to calculate the current result in other approaches too.

File: experiment/procedure.md [hunk 3]
```diff
@@ -1 +1,28 @@
-### Procedure
\ No newline at end of file
+## Demo
+* On the left hand side you can the order of the edges that the algorithm will follow.
```
Comment (raj-vlabs) [inline experiment/procedure.md]: Something missing in this sentence

File: experiment/procedure.md [hunk 4]
```diff
@@ -1 +1,28 @@
...
+* Current Edge and current iteration would be highlighted.
+* Youy can use next or previous button to move to next or previous state.
```
Comment (raj-vlabs) [inline experiment/procedure.md]: You misspelt as Youy. Please run your text pages through spell-checker

File: experiment/procedure.md [hunk 5]
```diff
@@ -1 +1,28 @@
...
+* For each iteration you need to select the edges that are going to be relaxed. You can select the edges by clicking on them.
+* After selecting the edges you can click on the Next Iteration button to move to the next iteration. You would be able to next iteration only if the edges that you have selected are correct.
+* You can move to previous iterations to see the changes in the distance and parent array.
+
+### Observation
+* The distance and parent array above the graph shows the distance from the source node to all the other nodes and the parent of each node respectively.
+* Obser the changes in the distance and parent array after each iteration.
+
+## Exercise
+* On the left hand side you can the order of the edges that the algorithm will follow.
```
Comment (raj-vlabs) [inline experiment/procedure.md]: Something missing in the sentence

File: experiment/procedure.md [hunk 6]
```diff
@@ -1 +1,28 @@
...
+* Select the iteration from the tabs above the graph and select the edges that are going to be relaxed. You can select the edges by clicking on them.
+* After selecting the edges in all iterations click on submit to check your answer.
```
Comment (raj-vlabs) [inline experiment/procedure.md:24]: I think we should also add that a selected edge can be deselected by clicking on it again

File: experiment/references.md [hunk 7]
```diff
@@ -1 +1,2 @@
-### Link your references in here
\ No newline at end of file
+1. https://en.wikipedia.org/wiki/Bellman%E2%80%93Ford_algorithm
```
Comment (raj-vlabs) [inline experiment/references.md]: Please use a descriptive text for the link and keep the url as a hyperlink so reader can get an idea what kind of link he is opening

File: experiment/simulation/css/main.css [hunk 8]
```diff
@@ -1 +1,189 @@
...
+  text-align: center;
+}
+
+.v-instruction-container {
+  margin-top: 30px;
+  border: .2rem solid #97d23d;
+  justify-content: flex-start;
+}
+
+.practiceMainContainer {
```
Comment (raj-vlabs) [inline experiment/simulation/css/main.css]: Use a consistent naming convention. For css selectors, use small case words separated by hyphens just like you have used above

File: experiment/simulation/demo.html [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    <title>Virtual labs</title>
+    <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/virtual-labs/virtual-style@0.0.8-b/css/style.min.css">
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/jsPlumb/2.15.6/js/jsplumb.min.js"></script>
+    <link rel="stylesheet" href="./css/main.css">
+</head>
+
+<body>
+    <div class="columns is-centered">
+        <div class="column">
+            <div class="v-collapsible-instruction-container v-instruction-container">
```
Comment (raj-vlabs) [inline experiment/simulation/demo.html]: Do we need both these classes?

File: experiment/theory.md [hunk 10]
```diff
@@ -1 +1,70 @@
...
+                    distance[v] = distance[u] + w
+    
+    In the above algorithm distance[v] represents the length of the shortest path from source to v.
+    They are initialized to infinity except for the source vertex, which is initialized to 0.
+    The outer loop is repeated V-1 times, where V is the number of vertices in the graph.
+    The inner loop relaxes all the edges in the graph.
+
+**Time Complexity**
+
+From the above pseudo code we can see that the outer loop is repeated V-1 times and the inner loop is repeated E times where E is the number of edges in the graph. So the time complexity of the algorithm is O(V*E). This is the average time complexity of the algorithm. In the worst case we would have a complete graph then E would be equal to V^2. So the time complexity of the algorithm in the worst case would be O(V^3). We can optimize the algorithm by stooping the outer loop if we have not updated the distance of any vertex in the inner loop. Thus our best case time complxity will be equal to O(E).
```
Comment (raj-vlabs) [inline experiment/theory.md]: We can optimize the algorithm by stopping the outer loop
in place of
We can optimize the algorithm by stooping the outer loop

File: experiment/theory.md [hunk 11]
```diff
@@ -1 +1,70 @@
...
+            for each edge (u, v) with weight w in Graph:
+                if distance[u] + w < distance[v]:
+                    distance[v] = distance[u] + w
+    
+    In the above algorithm distance[v] represents the length of the shortest path from source to v.
+    They are initialized to infinity except for the source vertex, which is initialized to 0.
+    The outer loop is repeated V-1 times, where V is the number of vertices in the graph.
+    The inner loop relaxes all the edges in the graph.
+
+**Time Complexity**
```
Comment (raj-vlabs) [inline experiment/theory.md:35]: Can we discuss the time complexity point-wise for each case, i.e. best, average, worst?

File: experiment/theory.md [hunk 12]
```diff
@@ -1 +1,70 @@
...
+
+From the above pseudo code we can see that the outer loop is repeated V-1 times and the inner loop is repeated E times where E is the number of edges in the graph. So the time complexity of the algorithm is O(V*E). This is the average time complexity of the algorithm. In the worst case we would have a complete graph then E would be equal to V^2. So the time complexity of the algorithm in the worst case would be O(V^3). We can optimize the algorithm by stooping the outer loop if we have not updated the distance of any vertex in the inner loop. Thus our best case time complxity will be equal to O(E).
+
+**Space Complexity**
+
+The space complexity of the algorithm is O(V) as we are using an array to store the distance of all the vertices from the source vertex. In the above pseudo code we are only calculating the shortest distance but the shortest path can be calculated by storing the parent of each vertex in an array. This will increase the space complexity of the algorithm to O(V*2) beacuse we are using two arrays to store the distance and parent of each vertex.
+
+
+**Pseudo Code for optimized algorithm with pareent array**
```
Comment (raj-vlabs) [inline experiment/theory.md]: Parent misspelled

File: experiment/theory.md [hunk 13]
```diff
@@ -1 +1,70 @@
...
+        for i = 1 to size of Graph:
+            for each edge (u, v) with weight w in Graph:
+                if distance[u] + w < distance[v]:
+                    distance[v] = distance[u] + w
+                    parent[v] = u
+
+
+**Negative Edges and Negative Cycle**
+
+Bellman-Ford is used over Dijkstra's Algorithm beacuse it handles garph with negative edges, but we need to ensure that the graph doesn't contain any negative cycle. In case of negative cycle the shortest distance will be minus infinity. For ex conider a cycle between edges A,B,C. The weights are
```
Comment (raj-vlabs) [inline experiment/theory.md]: Graph misspelled as garph

For example, consider a cycle
in place of
For ex conider a cycle

**Review by raj-vlabs — CHANGES_REQUESTED** [review]
1. Please run all content through spell-checker.
2. Please use type matching comparison operators throughout (=== and !==)
3. Bellman-Ford is not consistently used for the name. At least once in Theory page it is mentioned as Bellman ford

File: experiment/theory.md (hunk 11 above)
Comment (Aditya-debug15) [reply experiment/theory.md:35]: Updated the theory section to show them point wise

File: experiment/theory.md (hunk 12 above)
Comment (Aditya-debug15) [reply experiment/theory.md]: fixed it

File: experiment/theory.md (hunk 13 above)
Comment (Aditya-debug15) [reply experiment/theory.md]: corrected these mistakes

File: experiment/theory.md (hunk 10 above)
Comment (Aditya-debug15) [reply experiment/theory.md]: fixed the spelling mistake

File: experiment/procedure.md (hunk 5 above)
Comment (Aditya-debug15) [reply experiment/procedure.md]: changed it to `On the left-hand side you can see the order of the edges that the algorithm will follow`.

File: experiment/procedure.md (hunk 4 above)
Comment (Aditya-debug15) [reply experiment/procedure.md]: sorry for these kinds of mistakes i have run all the pages through a spell checker now

File: experiment/procedure.md (hunk 3 above)
Comment (Aditya-debug15) [reply experiment/procedure.md]: changed it to On the left-hand side you can see the order of the edges that the algorithm will follow.

File: experiment/procedure.md (hunk 6 above)
Comment (Aditya-debug15) [reply experiment/procedure.md:24]: I have updated it to `Select the iteration from the tabs above the graph and select the edges that are going to be relaxed. You can select the edges by clicking on them. To deselect an edge, click on it again.`

File: experiment/pretest.json (hunk 1 above)
Comment (Aditya-debug15) [reply experiment/pretest.json]: Modified the question

File: experiment/pretest.json (hunk 2 above)
Comment (Aditya-debug15) [reply experiment/pretest.json]: Modified the explanation

File: experiment/references.md (hunk 7 above)
Comment (Aditya-debug15) [reply experiment/references.md]: Added the descriptive text for links

File: experiment/simulation/demo.html (hunk 9 above)
Comment (Aditya-debug15) [reply experiment/simulation/demo.html]: `v-collapsible-instruction-container` class was not needed so i have removed it.

File: experiment/simulation/css/main.css (hunk 8 above)
Comment (Aditya-debug15) [reply experiment/simulation/css/main.css]: Updated the classes, now using hyphen separated naming convention

**Review by raj-vlabs — APPROVED** [review]

**Pull request merged**